Sources/GRPCCodeGen/Internal/Translator/IDLToStructuredSwiftTranslator.swift

```diff
@@ -112,87 +115,165 @@ extension IDLToStructuredSwiftTranslator {
+        if upperCaseNamespaces.contains(service.name.generatedUpperCase) {
           throw CodeGenError(
             code: .nonUniqueServiceName,
             message: """
-              Services with no namespace must not have the same names as the namespaces. \
-              \(service.name) is used as a name for a service with no namespace and a namespace.
+              Services with no namespace must not have the same generated upper case names as the namespaces. \
+              \(service.name.generatedUpperCase) is used as a generated upper case name for a service with no namespace and a namespace.
               """
           )
         }
       }
     }
 
-    // Check that service names are unique within each namespace.
-    for (namespace, services) in servicesByNamespace {
-      var serviceNames: Set<String> = []
+    // Check that the generated upper case names for services are unique within each namespace, to ensure that
+    // the service enums from each namespace enum have unique names.
+    for (namespace, services) in servicesByUpperCaseNamespace {
+      var upperCaseNames: Set<String> = []
+
       for service in services {
-        if serviceNames.contains(service.name) {
+        if upperCaseNames.contains(service.name.generatedUpperCase) {
           let errorMessage: String
           if namespace.isEmpty {
             errorMessage = """
-              Services in an empty namespace must have unique names. \
-              \(service.name) is used as a name for multiple services without namespaces.
+              Services in an empty namespace must have unique generated upper case names. \
+              \(service.name.generatedUpperCase) is used as a generated upper case name for multiple services without namespaces.
               """
           } else {
             errorMessage = """
-              Services within the same namespace must have unique names. \
-              \(service.name) is used as a name for multiple services in the \(service.namespace) namespace.
+              Services within the same namespace must have unique generated upper case names. \
+              \(service.name.generatedUpperCase) is used as a generated upper case name for multiple services in the \(service.namespace.base) namespace.
               """
           }
           throw CodeGenError(
             code: .nonUniqueServiceName,
             message: errorMessage
           )
         }
-        serviceNames.insert(service.name)
+        upperCaseNames.insert(service.name.generatedUpperCase)
       }
     }
   }
 
-  // Verify method names are unique for the service.
+  // Verify method names are unique within a service.
   private func checkMethodNamesAreUnique(
     in service: CodeGenerationRequest.ServiceDescriptor
   ) throws {
-    let methodNames = service.methods.map { $0.name }
-    var seenNames = Set<String>()
+    // Check that the method descriptors are unique, by checking that the base names
```

Comment:
This function does the same thing three times but with different information. Can you refactor this to avoid the duplication?